Add OpenSpiel interface

jhtschultz opened the pull request:
This pull request is a follow up to https://github.com/deepmind/acme/issues/80 and adds the ability to run Acme’s DQN on OpenSpiel games. All changes and additions are contained in the acme/open_spiel directory. The example `run_dqn.py` runs Acme’s DQN on OpenSpiel’s tic-tac-toe. It’s set up to print out a full playthrough (policies included) every 1000 episodes making it easy to verify that the agent is in fact learning a good policy. The agents start playing sensibly after roughly 10k episodes, and by about 50k episodes pretty much every game in which an exploratory action is not taken results in a draw.

Though I’ve endeavored to keep the code clean and production ready, this PR is only intended to be a starting point. There are a number of TODOs. Some are only meant to call attention to important considerations, others are tasks which seemed better addressed after there’s been a chance for feedback and discussion.

Perhaps the most important design feature, and the one that stands to improve most from further thought, is the environment wrapper. In an effort to keep as much existing structure intact as possible, the wrapper is built around OpenSpiel’s python `rl_environement` - which notably is not a `dm_env`. An alternative approach is to wrap the OpenSpiel game directly as a `dm_env`. However, the `dm_env` doesn’t appear to have been explicitly designed with multiplayer games in mind. Perhaps someone who has worked on `dm_env`, or the related (and very useful) `tree` repo might have some insight into how best to do this.

I look forward to feedback and I’ll be happy to implement any suggested changes. You’ll notice that very few modifications to the DQN agent/learner were necessary. With this interface it should be straightforward to adapt other Acme algorithms to run on OpenSpiel games :-)

jhtschultz commented:
And to import OpenSpiel, add these two lines to Acme's virtual environment.

```
export PYTHONPATH=$PYTHONPATH:/<path_to_open_spiel>
export PYTHONPATH=$PYTHONPATH:/<path_to_open_spiel>/build/python
```

mwhoffman commented on acme/open_spiel/agents/tf/dqn/learning.py:
```diff
@@ -0,0 +1,211 @@
+    with tf.GradientTape() as tape:
+      # Evaluate our networks.
+      q_tm1 = self._network(o_tm1)
+      q_t_value = self._target_network(o_t)
```
Is this the main change for DQN in this case? IE I wonder (as I think @fastturtle was alluding to) if the `legal_actions` mask is moved into the observations directly if this logic can then be moved into the policy (perhaps via a wrapper around a "raw q-function network"). This would then allow a broader array of potential algorithms to be applied in this setting.

mwhoffman commented:
Take this all with a grain of salt, because I'm thinking out loud a little bit at the moment, but:

I don't know how much of a pain it would be to retroactively mark some of these files as copies (i.e. `git cp`). I'm thinking _mostly_ with respect to the learner, but as I think this might be painful, could you highlight some of the changes made here just to center the discussion?

But from looking I think the main changes in the agent/actor/learner (please correct me if I'm wrong!) revolve around the legal actions. I've noted this above, and I wonder if we can push these directly as observations and let the network handle the legal actions. I think this will allow the learner to remain unchanged. Ideally the same could be done for actor/agent (putting aside the changes to epsilon greedy that are probably needed). But if this only really helps the learner that'd be great.

Ideally we could have something where we have an openspiel-specific Environment Loop, wrapper, and potentially a "Q-network wrapper" but the agent itself isn't changed (and we can peel back as much of that as is necessary if/when this ideal isn't possible 😄).

jhtschultz commented:
Ahh I see, yeah moving the legal actions mask into the observation and letting the network handle everything sounds like the right approach. Should streamline things quite a bit.

Tried to mark the new files as copies but couldn’t get the git diff to work, it kept marking everything as new additions. Either way I think you already identified the core issue. I’ll implement those changes with the goal of confining everything to the network, loop and wrapper. Thanks for the quick feedback!

jhtschultz commented:
Pushed a big update that's pretty close the ideal @mwhoffman described. Moving legal actions into the observation really was the way to go, simplified everything dramatically. All that’s needed now to run the OpenSpiel example is the wrapper, environment loop, custom networks, and the slightest of modifications to the DQN agent. 

The issue with the DQN agent is the policy network. Currently the DQN agent creates the policy network by appending trfl’s `epsilon_greedy()` layer to the learning network. As far as I can tell, no matter what our custom network outputs this layer will select illegal actions. My workaround was to add the policy network as an optional argument. It offers more flexibility while not changing the default behavior of the DQN agent.

Now that OpenSpiel’s footprint is much smaller, I moved it out of its own directory and placed the files in the appropriate folders. The wrapper is < 150 lines, env loop < 250, and the networks are pretty simple, so this update gets us a lot closer to the finish line.

fastturtle commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,247 @@
+  def __init__(
+      self,
+      environment: dm_env.Environment,
+      actors: List[core.Actor],
```
This could be Sequence[core.Actor] right?

jhtschultz replied:
Fixed

fastturtle commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,247 @@
+                           discount=timestep.discount[player],
+                           step_type=timestep.step_type)
+
+  # TODO Remove? Currently used for debugging.
```
I'm not very familiar with TODOs in OS code, but should we have an issue to track them?

jhtschultz replied:
I'll clear out all the TODOs before it gets merged, there aren't many. Left these here just to make sure they got addressed.

fastturtle commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,247 @@
+                           step_type=timestep.step_type)
+
+  # TODO Remove? Currently used for debugging.
+  def _print_policy(self, timestep: dm_env.TimeStep, player: int):
```
Is this generally useful, or was it while you were debugging a particular issue?

jhtschultz replied:
This is here so that simply by running the environment loop one can easily verify the agents are learning correctly.

I have mixed feelings as to whether it should make the final cut. It's nice for the user to be able to run an example and see exactly what's going on without any additional work on their part. On the other hand, it's of questionable usefulness in larger environments and makes the environment loop slightly less elegant. If anyone has a preference I'm happy to go with that.

fastturtle commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,247 @@
+
+  def __init__(
+      self,
+      environment: dm_env.Environment,
```
This requires a OpenSpielWrapper.

jhtschultz replied:
Fixed

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,145 @@
+    """Resets the episode."""
+    self._reset_next_step = False
+    open_spiel_timestep = self._environment.reset()
+    assert open_spiel_timestep.step_type == rl_environment.StepType.FIRST
```
Can you replace asserts with if/raise. That way we can give more informative error messages.

jhtschultz replied:
Replaced all asserts with if/raise. This one in particular can be safely removed, calling `self._environment.reset()` has no choice but to return `StepType.FIRST`. The gym and Atari wrappers don't do a check after `reset()` either.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,145 @@
+    elif step_type == rl_environment.StepType.LAST:
+      step_type = dm_env.StepType.LAST
+    else:
+      raise ValueError("Did not recognize OpenSpiel StepType")
```
Print the type it didn't recognize for debugging.

jhtschultz replied:
Fixed

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,145 @@
+  # Convert OpenSpiel observation so it's dm_env compatible. Also, the list
+  # of legal actions must be converted to a legal actions mask.
+  def _convert_observation(
+      self, open_spiel_timestep: NamedTuple) -> types.NestedArray:
```
Is there a more specific type, rather than NamedTuple?

jhtschultz replied:
Yes, updated to `rl_environment.TimeStep`.

fastturtle commented on acme/tf/networks/legal_actions.py:
```diff
@@ -0,0 +1,117 @@
+    super(MaskedSequential, self).__init__(name=name)
+    self._layers = list(layers) if layers is not None else []
+    self._illegal_action_penalty = -1e9
+    # TODO Note: illegal_action_penalty cannot be -np.inf, throws
```
Turn this TODO into a Note?

jhtschultz replied:
Fixed

fastturtle commented on acme/tf/networks/legal_actions.py:
```diff
@@ -0,0 +1,117 @@
+    return outputs
+
+
+# TODO Function to update epsilon
```
What does this TODO refer to?

jhtschultz replied:
Ah good question, I meant to ask...is there currently an easy way to decay DQN's epsilon parameter? There doesn't appear to be, it seems fixed once the policy network is created. This is an important feature for some use cases and since I was building an epsilon greedy sonnet module anyway, it seemed like something we'd want to include.

jhtschultz replied:
Fixed.

fastturtle commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,247 @@
+  return np.zeros(spec.shape, spec.dtype)
+
+
+# Internal class.
```
Extraneous comment?

jhtschultz replied:
Fixed

fastturtle reviewed (changes requested):
Thanks for the PR, left some comments :)

fastturtle commented:
Also there's an issue with pytest failing, looks like it isn't installing openspiel.

jhtschultz commented:
@fastturtle thanks for the comments! Regarding pytest failing, OpenSpiel does not currently have a pip package for direct installation (though it's in the works). Given that, any advice on how to handle this? And tagging @lanctot, who knows all things about installing OpenSpiel :)

lanctot commented:
@fastturtle  @jhtschultz  I am planning for an OpenSpiel release in a few weeks, and have it in PyPi, but only by source distribution for now (which means it'll be 1-2 min install, and requires the machines have CMake, clang++, and python3-dev). Do you think that would solve it?

jhtschultz commented:
Hi! Wanted to check in on the status of this PR. 

I saw commit https://github.com/deepmind/acme/commit/996c63db160a3daec376ec8fe5ed3b1b3a3d16c1 followed the approach taken here and added the ability to provide DQN's policy network directly. That's great because now this PR requires zero changes to Acme’s DQN. 

Commit https://github.com/deepmind/acme/commit/996c63db160a3daec376ec8fe5ed3b1b3a3d16c1 also added `NetworkWithMaskedEpsilonGreedy`, which uses `trfl.epsilon_greedy()`. As documented here https://github.com/deepmind/trfl/issues/27 and https://github.com/deepmind/trfl/pull/28, there is a bug in trfl’s action masking which currently renders this network unusable. Once fixed, I could refactor the networks used in this PR to use `NetworkWithMaskedEpsilonGreedy` instead.

I’ve been implementing NFSP, a deep-RL algorithm for imperfect information games, using this Acme-OpenSpiel interface and the interface works quite well. If we can identify any remaining obstacles to importing this PR I'll be happy to work on them. Thanks!

lanctot commented:
> @fastturtle thanks for the comments! Regarding pytest failing, OpenSpiel does not currently have a pip package for direct installation (though it's in the works). Given that, any advice on how to handle this? And tagging @lanctot, who knows all things about installing OpenSpiel :)

I spoke to @fastturtle yesterday. OpenSpiel does have a pip now but it still requires a build from source, which requires specific things from your environment and could take a long time. So, until we have a binary wheel (at least a few months), we've agreed just disable tests that require/depend on OpenSpiel. 

Please keep the code for these tests around, though. When it's merged, I will enable this test internally to ensure that it passes on our side. Thanks!

jhtschultz commented:
Alright all checks are passing! And I added two simple tests for the wrapper and environment loop. Following the approach used for other Acme tests, they are skipped when they run into an import error, so they don’t need to be disabled/reenabled @lanctot.

Small detail: I tried to follow the pattern used throughout Acme with `__init__.py` imports, but couldn’t avoid pytest failing if OpenSpiel imports were included in `__init__.py`. For the time being they’re commented out. Maybe there’s some other work around, perhaps it’s best to delete them altogether.

fastturtle commented on acme/tf/networks/legal_actions.py:
```diff
@@ -0,0 +1,119 @@
+    super(MaskedSequential, self).__init__(name=name)
+    self._layers = list(layers) if layers is not None else []
+    self._illegal_action_penalty = -1e9
+    # Note: illegal_action_penalty cannot be -np.inf, throws error "Priority
```
This is interesting, did this surprise you? Is the illegal action being used to calculate the priority somehow?

jhtschultz replied:
Got to the bottom of this. Trfl’s `qlearning` ops use a `batched_index` function to return the action value of selected actions: `qa_tm1 = indexing_ops.batched_index(q_tm1, a_tm1)`

And `batched_index` computes: `tf.reduce_sum(values * one_hot_indices, axis=-1, keepdims=keepdims)`

This returns NaN whenever `-np.inf` is present among the actions values. These NaNs show up in the td_error, which then gets passed to the priorities, which throws the error. 

Note: this examination prompted the following commit fa986b20be5244dcc72b49bcbda50dbe4adf0cf3 allowing the MaskedSequential module to handle `-np.inf` correctly. My first version of MaskedSequential did handle `-np.inf` correctly, but ran into the priority error, so I switched to using `-1e9`. Then I subsequently added this line of code: `outputs = outputs * (1 - terminal)` which seemed cleaner than what was previously there, but broke the ability to handle `-np.inf`. Either way, whether `-np.inf` is used or not, I think the update in this commit is the correct way to force the network to output all zeros in a terminal state.

**TLDR**: Ultimately, outputting `-np.inf` for illegal actions is incompatible with trfl. There was also an issue with how MaskedSequential handled `-np.inf` which is now fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+      self, open_spiel_timestep: rl_environment.TimeStep) -> types.NestedArray:
+    observations = []
+    for pid in range(self._environment.num_players):
+      legals = np.zeros(self._environment._game.num_distinct_actions())
```
Is there any way to get this information without accessing a private attribute? Could it be exposed from within the rl_environment.Environment?

jhtschultz replied:
Yes, not sure how I overlooked these. In all but one case (`environment._use_observation`) the attribute is already accessible as a property from within `rl_environment.Enviroment` and these have been fixed.

I'll see about adding `use_observation` as a property to `rl_environment.Environment`. For now it's the only private attribute we still need to access.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+          open_spiel_timestep.observations["info_state"][pid]),
+                               legal_actions=legals,
+                               terminal=np.asarray(
+                                   [float(open_spiel_timestep.last())]))
```
A word of caution: numpy will convert this into a np.float64 if you don't explicitly specify the dtype.

jhtschultz replied:
Good call, fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+  def observation_spec(self) -> types.NestedSpec:
+    # Observation spec depends on whether the OpenSpiel environment is using
+    # observation/information_state tensors
+    if self._environment._use_observation:
```
Same comment about private attributes.

jhtschultz replied:
See comment above.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+    # observation/information_state tensors
+    if self._environment._use_observation:
+      return OLT(observation=specs.Array(
+          (self._environment._game.observation_tensor_size(),), np.float32),
```
Same comment about private attributes.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+      return OLT(observation=specs.Array(
+          (self._environment._game.observation_tensor_size(),), np.float32),
+                 legal_actions=specs.Array(
+                     (self._environment._game.num_distinct_actions(),),
```
Same comment about private attributes.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+                 terminal=specs.Array((1,), np.float32))
+    else:
+      return OLT(observation=specs.Array(
+          (self._environment._game.information_state_tensor_size(),),
```
Same comment about private attributes.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+          (self._environment._game.information_state_tensor_size(),),
+          np.float32),
+                 legal_actions=specs.Array(
+                     (self._environment._game.num_distinct_actions(),),
```
Same comment about private attributes.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+
+  @property
+  def current_player(self) -> int:
+    return self._environment._state.current_player()
```
Same comment about private attributes.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,144 @@
+
+from typing import NamedTuple
+
+# pytype: disable=import-error
```
Can you narrow this down (and re-enable it).

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper_test.py:
```diff
@@ -0,0 +1,67 @@
+
+try:
+  # pylint: disable=g-import-not-at-top
+  # pytype: disable=import-error
```
Re-enable.

jhtschultz replied:
Fixed.

fastturtle commented on acme/wrappers/open_spiel_wrapper_test.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@unittest.skipIf(SKIP_OPEN_SPIEL_TESTS, SKIP_OPEN_SPIEL_MESSAGE)
+class OpenSpielWrapperTest(parameterized.TestCase):
```
The test cases aren't parameterized.

jhtschultz replied:
Fixed.

fastturtle commented on examples/open_spiel/run_dqn.py:
```diff
@@ -0,0 +1,75 @@
+  # Construct the agents.
+  agents = []
+
+  for i in range(environment.num_players):
```
Nit: It's a bit more pythonic to do:

```
for network, policy_network in zip(networks, policy_networks):
  ...
```

jhtschultz replied:
Yep that's much better. Nitty comments are always appreciated :) Fixed.

fastturtle commented on examples/open_spiel/run_dqn.py:
```diff
@@ -0,0 +1,75 @@
+  # Run the environment loop.
+  loop = open_spiel_environment_loop.OpenSpielEnvironmentLoop(
+      environment, agents)
+  loop.run(num_episodes=100000)  # pytype: disable=attribute-error
```
This shouldn't raise an attribute error, any idea whats going on?

jhtschultz replied:
Silly code artifact from using the bsuite run_dqn.py example as a template. Fixed.

fastturtle reviewed (commented):
Apologies for the long turn around, we're all eager to get this in!

jhtschultz commented:
@fastturtle thanks for looking it over! Most of the changes are already in. I still need to do some further digging to fully explain the `-np.inf` threshold causing the priority error. Good call on your part to highlight that. I'll address the comments individually and let you know when everything is ready to go on my end.

fastturtle commented:
@jhtschultz sounds good, I'll keep an eye out for your comments.

jhtschultz commented on acme/environment_loops/__init__.py:
```diff
@@ -0,0 +1,17 @@
+
+"""Specialized environment loops."""
+
+#from acme.environment_loops.open_spiel_environment_loop import OpenSpielEnvironmentLoop
```
That's right, and I agree the try-except block is the way to go. Fixed.

jhtschultz commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,222 @@
+import time
+from typing import Optional, Sequence
+
+# pytype: disable=import-error
```
Fixed.

jhtschultz commented on acme/environment_loops/open_spiel_environment_loop_test.py:
```diff
@@ -0,0 +1,102 @@
+
+try:
+  # pylint: disable=g-import-not-at-top
+  # pytype: disable=import-error
```
Fixed.

jhtschultz commented on acme/tf/networks/__init__.py:
```diff
@@ -34,6 +34,8 @@
 from acme.tf.networks.distributional import UnivariateGaussianMixture
 from acme.tf.networks.distributions import DiscreteValuedDistribution
 from acme.tf.networks.duelling import DuellingMLP
+#from acme.tf.networks.legal_actions import MaskedSequential
```
Fixed.

jhtschultz commented:
@fastturtle Done! Not sure why some pytests get stuck. Recent commits, including the previous one https://github.com/deepmind/acme/pull/87/commits/ff9b7addd5ee83c333f5c0725672673446f01477, have been consistently passing so I think everything is good to go. Let me know if there's anything else I can do.

jhtschultz commented:
We exposed `_use_observation` as a property in OpenSpiel's rl_environment, and the wrapper here has been updated to use the property.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+      observations.append(player_observation)
+    return observations
+
+  def observation_spec(self) -> types.NestedSpec:
```
This actually returns a OLT.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+                     np.float32),
+                 terminal=specs.Array((1,), np.float32))
+
+  def action_spec(self) -> types.NestedSpec:
```
Returns a specs.DiscreteArray.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+  def action_spec(self) -> types.NestedSpec:
+    return specs.DiscreteArray(self._environment.game.num_distinct_actions())
+
+  def reward_spec(self) -> types.NestedSpec:
```
Returns a specs.BoundedArray.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+                              minimum=self._environment.game.min_utility(),
+                              maximum=self._environment.game.max_utility())
+
+  def discount_spec(self) -> types.NestedSpec:
```
Returns a specs.BoundedArray.

fastturtle commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+  # Convert OpenSpiel observation so it's dm_env compatible. Also, the list
+  # of legal actions must be converted to a legal actions mask.
+  def _convert_observation(
+      self, open_spiel_timestep: rl_environment.TimeStep) -> types.NestedArray:
```
Returns a List[OLT].

fastturtle reviewed (commented):
Looks really good, some last comments from me 👍

jhtschultz commented:
Thanks! Good catches, fixed 👍

fastturtle reviewed (commented):
This looks good to me, and has my approval. @mwhoffman PTAL.

mwhoffman commented on acme/wrappers/open_spiel_wrapper.py:
```diff
@@ -0,0 +1,147 @@
+# pytype: enable=import-error
+
+
+class OLT(NamedTuple):
```
It might make more sense to define the OLT class outside this wrapper (or in the legal_actions policy definitions) and import here. That way we can skip one level of trying and potentially use the legal actions policies outside of openspiel envs. But we can play around with that after getting this change in...

qstanczyk commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,221 @@
+    # Run an episode.
+    while not timestep.last():
+      # Generate an action from the agent's policy and step the environment.
+      if self._environment.is_turn_based:
```
Hi, I was thinking about possibility of making this environment loop more generic by covering cases of multi-agent/multi-team environments that are not necessarily turn based or fully simultaneous.
Lets consider multi-agent, turn-based environment (for example soccer environment like https://github.com/google-research/football/) where each team consists of 11 players, but teams execute actions in turns. To cover OpenSpiel and this use case with one implementation we could:
- eliminate _environment.is_turn_based
- replace _environment.get_state.current_player() with a set of current players
That would cover "# TODO Support simultaneous move games." in a generic way.
What do you think about that?

lanctot replied:
Hi @qstanczyk,

Can you elaborate on your motivation for this? Why is this a good thing? I worry because if we make this too general, it gets harder to interface with the specific API it was designed to be used with and could introduce some awkward bugs for use cases we didn't anticipate by supporting multiple APIs.

Also I don't understand the football env example. If that environment is doing turn-based already, then it should fit exactly the OpenSpiel API as well already (which as you say is already turn-based and you just have 11 consecutive actions from the same player before switching the current player to the next one)? 

WDYT about a wrapper of the football env as an OpenSpiel python game (living in this repos)? Sounds like they are compatible from your description. Let me know if I'm missing something.

qstanczyk replied:
If turn-based Soccer environment would be integrated the way you suggest (11 actions for each team) then what is the use case you anticipate for "!is_turn_based"? Any multi-agent environment can be implemented by passing N actions, so maybe in such case there is no need to have "is_turn_based" check around?
My motivation was to try and generalize the API by supporting any is_turn_based / multi-agent mix. I'm fine with not taking this thought into account, but I just wonder at this point if "is_turn_based" is needed.

lanctot replied:
> If turn-based Soccer environment would be integrated the way you suggest (11 actions for each team) then what is the use case you anticipate for "!is_turn_based"?

Wait, is that not how it's implemented? I only assumed that based on what you said. The !is_turn_based is necessary here because of the way OpenSpiel categorizes its dynamics. Some games require actions from all agents (simultaneous game) and apply them all at once, and others are turn-based so you only get one for the current player.

> Any multi-agent environment can be implemented by passing N actions, so maybe in such case there is no need to have "is_turn_based" check around?

Yes, I agree, but that's not how OpenSpiel implements turn-based games.

> My motivation was to try and generalize the API by supporting any is_turn_based / multi-agent mix.

Ok yes I think I understand. I think this would make more sense if we had a use case for this generalization, and I'm not seeing it (yet). So my hesitatation is due mainly to this wrapper being designed for OpenSpiel specifically.

OpenSpiel treats these two cases separately because the algorithms that are built on top of the different type of games depend on these choices. Also, in OpenSpiel a simultaneous game doesn't need to be fully simultaneous at all states and also it's fine for a subset of agents to only have 1 legal pass-like move, hence my suggestion of wrapping the ones you mention as OpenSpiel simultaneous-move games, because then this interface would work on them.

qstanczyk replied:
Ack, I was just thinking out-loud about generalization of the API. We don't look at hooking up GRF at this point, but was wondering if this OpenSpiel environment loop could be more generic.

qstanczyk commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,221 @@
+          if self._should_update:
+            self._actors[player_id].update()
+      self._observed_first = [False] * len(self._actors)
+      self._prev_actions = [pyspiel.INVALID_ACTION] * len(self._actors)
```
Maybe initialize these two once at the beginning of run_episode? That would eliminate code duplication.

qstanczyk commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,221 @@
+      if not self._observed_first[player]:
+        player_timestep = dm_env.TimeStep(
+            observation=timestep.observation[player],
+            reward=None,
```
Doesn't OpenSpiel return reward in this case? In generic case I think we should use reward provided by the environment. Can't we call _get_player_timestep instead (setting dm_env.StepType.FIRST)?

qstanczyk commented on acme/environment_loops/open_spiel_environment_loop.py:
```diff
@@ -0,0 +1,221 @@
+
+    # Make the first observation.
+    self._send_observation(timestep, self._environment.current_player)
+
```
Can't _send_observation be called as the first thing in the loop (which would eliminate duplicated code)?

mwhoffman commented:
A few minor changes were made (mostly linter reasons!) and this was submitted in https://github.com/deepmind/acme/commit/7fa9b804e47fd1b1fbb20ed5e62275d822607c8b. So I'll close this down for now, and any further discussion can carry on outside this PR. Thanks for all the hard work!
